java/ql/src/semmle/code/java/dataflow/FlowSources.qll

```diff
@@ -318,3 +319,24 @@ class AndroidIntentInput extends DataFlow::Node {
     )
   }
 }
+
+/** Exported Android `Intent` that may have come from a hostile application. */
+class ExportedAndroidIntentInput extends RemoteFlowSource {
```

Comment:
Don't duplicate the logic of `AndroidIntentInput`. How about something like


```codeql
/** Android `Intent` that may have come from a hostile application. */
class AndroidIntentInput extends DataFlow::Node {
  Type receiverType

  AndroidIntentInput() {
    exists(MethodAccess ma, AndroidGetIntentMethod m |
      ma.getMethod().overrides*(m) and
      this.asExpr() = ma and
      receiverType = ma.getMethod().getDeclaringType()
    )
    or
    exists(Method m, AndroidReceiveIntentMethod rI |
      m.overrides*(rI) and
      this.asParameter() = m.getParameter(1) and
      receiverType = m.getDeclaringType()
    )
  }
}

/** Exported Android `Intent` that may have come from a hostile application. */
class ExportedAndroidIntentInput extends RemoteFlowSource, AndroidIntentInput {
  ExportedAndroidIntentInput() {
    receiverType.(ExportableAndroidComponent).isExported()
  }

  override string getSourceType() { result = "Exported Android intent source" }
}
```

Reply:
Thanks for the code. I've changed the implementation so that it follows proper class inheritance and is much cleaner.